### PR TITLE
Lib control rewrite

### PR DESCRIPTION
**Pull request by oorzkws**
You're probably sick of me talking about it at this point, here's the final(pending review) changes.

### CODE REVIEW

**Comment by oorzkws** [comment]
Requesting review, @Pezzawinkle @Samuel2507 @SqFKYo

File: omnilib/control.lua [hunk 1]
```diff
@@ -1,408 +1,343 @@
...
+                                    -- RIP MT. OMNISSIAH
+                                --
+                            --
+                        --
+                    --
+                --
+            --
+        --
+    --
+--
```
Comment (oorzkws) [inline omnilib/control.lua:20]: To be clear, this was requested by Zelos.

File: omnilib/control.lua [hunk 2]
```diff
@@ -1,408 +1,343 @@
...
+		elseif not global.omni.need_update then -- If we need_update we'll be iterating everything anyway.
+			update_tech(tech, true)
+			if tech.name:find("^omnitech") and not tech.name:find("^omnipressed") then
+				global.omni.update_buildings = true
 			end
 		end
 	end
 end)
 
+script.on_event(defines.events.on_technology_effects_reset, function(event)
```
Comment (oorzkws) [inline omnilib/control.lua:322]: I think this has a possibility of creating an infinite loop, if something locks a tech we'd normally unlock and then resets tech effects, feedback requested.

File: omnimatter_compression/control.lua [hunk 3]
```diff
@@ -46,54 +46,6 @@ script.on_event("decompress-stack", function(event)
...
-		elseif tech.force.technologies["compression-"..kind.."-buildings"].researched then
-			for _,eff in pairs(tech.effects) do
-				if eff.type == "unlock-recipe" and not_already_compressed(eff.recipe) and tech.force.recipes[eff.recipe.."-compressed-"..kind] then
-					tech.force.recipes[eff.recipe.."-compressed-"..kind].enabled = true
-				end
-			end
-		end
-	end
-end)
-
```
Comment (oorzkws) [inline omnimatter_compression/control.lua:96]: Removed as functionality is now present in lib/control

File: omnilib/control.lua [hunk 4]
```diff
@@ -1,408 +1,343 @@
...
+					base = base, -- The entry name in the common table
+					tier = tier,
+					variant = variant,
+					compressed = (index == 3)
+				}
+				recipe_map[base] = recipe_map[base] or {}
+				recipe_map[base].active_tier = math.min(97, recipe_map[base].active_tier or tier)
+				recipe_map[base][tier] = recipe_map[base][tier] or {}				
+				recipe_map[base][tier][variant] = name
+				break
```
Comment (oorzkws) [inline omnilib/control.lua:271]: A lot is happening here, so I'll explain:
We check every recipe against our recipe-name-tierletter-modifier pattern. The first pattern indicates a recipe has no modifiers, AKA it is not compressed or permuted. The second is for permutations, and the third will indicate if the recipe is compressed.

From here, our data scheme has two parts. We have recipe_map, which keeps our current state, and recipes which keeps the keys and metadata we use to lookup items in recipe_map.

Recipes are indexed by prefix, then by tier as the letter in bytes, then by suffix. An example dump from just omnimatter/omnipermute is below.

https://gist.github.com/oorzkws/67f0599a7968c3b3ced91e689d8b37ff

File: omnilib/control.lua [hunk 5]
```diff
@@ -1,408 +1,343 @@
...
+	"nanite",
+	"quantum",
+	"singularity"
+}
+
+local function update_tech(tech)
+	local variant = tech.force.technologies["omnipressed-" .. tech.name] or tech.force.technologies[tech.name:gsub("^omnipressed%-", "")] or {}
+	-- Handle compressed techs
+	if tech.researched or variant.researched then
+		tech.researched, variant.researched = true, true
```
Comment (Pezzawinkle) [inline omnilib/control.lua:131]: how is this not crashing on multi-level research, like infinite tech? the version that was in the compression control file specifically asked for level for a reason ;)

Comment (oorzkws) [reply omnilib/control.lua]: 7efa654 should address your concerns. It didn't crash but it wasn't properly syncing them.

**Comment by oorzkws** [comment]
Looking into some possible issues with building unlocks

File: omnilib/control.lua (hunk 1 above)
Comment (Pezzawinkle) [reply omnilib/control.lua:20]: As it should be, leave some comedy in the code :D

File: omnilib/control.lua [hunk 6]
```diff
@@ -1,408 +1,420 @@
...
+	end
+	full_refresh = full_refresh or global.omni.need_update
+	local recipe_tree, metadata = get_recipe_family(recipe)
+	if not recipe_tree then
+		return
+	end
+	-- We're the top of the tree!
+	if recipe_tree and metadata.variant == "" and recipe_tree.active_tier < metadata.tier then
+		local I = full_refresh and 97 or recipe_tree.active_tier
+		local research_status = not not (recipe.force.technologies["compression-recipes"] or {}).researched
```
Comment (Samuel2507) [inline omnilib/control.lua:46]: not not?

File: omnilib/controlFunctions.lua [hunk 7]
```diff
@@ -12,13 +13,6 @@ function split(inputstr, sep)
...
-function invOrd(l)
-	for j,i in pairs(ord) do
-		if l==i then return j end
-	end
-	return nil
-end
-
 function start_with(a,b)
 	return string.sub(a,1,string.len(b)) == b
 end
```
Comment (Samuel2507) [inline omnilib/controlFunctions.lua:18]: as we agreed on discord, this should be renamed into "control-functions"

File: omnilib/control.lua (hunk 6 above)
Comment (oorzkws) [reply omnilib/control.lua:46]: "not not" is equivalent in usage to a boolean cast

File: omnilib/control.lua [hunk 8]
```diff
@@ -1,408 +1,420 @@
...
-							z=z+1
+
+local building_tiers = {
+	compact = "compression-compact-buildings",
+	nanite = "compression-nanite-buildings",
+	quantum = "compression-quantum-buildings",
+	singularity = "compression-singularity-buildings"
+}
+
+local function update_tech(tech)
```
Comment (Pezzawinkle) [inline omnilib/control.lua:129]: something is not working here, can you please check that as a compressed tech unlocks, it also gets the paired one done, and also the other way, it is not functioning on my test map. (It works well for multi-tiered and infinite tech, but is failing on regular tech like angels smelting and omnitraction)

Comment (oorzkws) [reply omnilib/control.lua:129]: Fixed below along with one other possible loop as well.

**Pull request merged**